### PR TITLE
Add RTDETR Trainer

### PR DESCRIPTION
**Pull request by AyushExel**
- [x] Port Base DETRLoss class
 - [x] Port DinoLoss Class
 - [x] Build RTDETR trainer
 - [x] Verify training

### Summary
🚀🛠️🗑️

This pull request adds a new loss function for the DETR model, which is a transformer-based object detection and segmentation model. The pull request modifies the `ultralytics/yolo/utils/loss.py` file to use the new loss function, and removes the `FocalLoss` class from the `ultralytics/yolo/utils/metrics.py` file to avoid duplication. The pull request also adds a new class `DETRLoss` to the `ultralytics/vit/utils/loss.py` file, which implements the loss function for the DETR model.

> _`DETRLoss` added_
> _Transformer for detection_
> _Autumn of convnets_

### Walkthrough
*  Add a new loss function for the DETR model, which is a transformer-based object detection and segmentation model ([link](https://github.com/ultralytics/ultralytics/pull/2745/files?diff=unified&w=0#diff-23d48c8beef123f219ad06db989623001af8745b723eb4dc06b264d50c062eeeR1-R346)). The new loss function is implemented in the `DETRLoss` class in the `ultralytics/vit/utils/loss.py` file, and it inherits from `nn.Module`. The class takes several arguments to configure the loss coefficients, the matcher, the auxiliary loss, and the focal loss options. The class defines methods to compute the class, bbox, mask, and dice losses, as well as the auxiliary losses for each decoder layer. The class also defines a `forward` method that takes the predictions and the targets and returns a dictionary of losses.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces the ability to train RT-DETR models, updates focal and varifocal loss implementations, and includes several other improvements and bug fixes.

### 📊 Key Changes
- Enabled training for RT-DETR models with `model.train()` method.
- FocalLoss and VarifocalLoss classes added/updated in loss modules.
- Renamed and moved some `.yaml` files related to models.
- Adjusted tests to differentiate between YOLO and RTDETR models.
- Updated citation keywords in documentation.
- Fixed condition checks and logic in various modules like SAM and dataset augmentation.

### 🎯 Purpose & Impact
- Users can now train RT-DETR models for object detection, facilitating the use of Transformer-based architectures.
- FocalLoss functionalities provide better focus on hard-to-classify examples, which may improve model accuracy and convergence speed.
- The correct handling and differentiation between model types in tests prevent potential errors and ensure consistency.
- Documentation updates provide accurate information and increase the findability of relevant resources.
- Bug fixes and condition adjustments make the codebase more robust and reliable, reducing unexpected errors.

### CODE REVIEW

**Comment by Laughing-q** [comment]
@AyushExel I updated some api from paddle to torch, like `paddle.concat` -> `torch.cat`, `paddle.to_tensor` -> `torch.tensor`. And left a comment with TODO to these api I haven't verified.

**Comment by AyushExel** [comment]
@Laughing-q okay got it. Did you figure out the gt_meta part ?

**Comment by AyushExel** [comment]
> @AyushExel I updated some api from paddle to torch, like `paddle.concat` -> `torch.cat`, `paddle.to_tensor` -> `torch.tensor`. And left a comment with TODO to these api I haven't verified.

yeah up until now I've just replaced paddle with torch and ran until where it errored out. The execution reached till gt_meta, after that I'm stuck

**Comment by Laughing-q** [comment]
@AyushExel I took a look at their code. I believe it's a dict with several elements: `gt_bbox`, `gt_class`. And the bbox format is `xyxy`. 
[https://github.com/PaddlePaddle/PaddleDetection/blob/eeebef9f3f8347eb18f6b55cf45c2ed38221c3f6/ppdet/data/source/coco.py#L155-L163](https://github.com/PaddlePaddle/PaddleDetection/blob/eeebef9f3f8347eb18f6b55cf45c2ed38221c3f6/ppdet/data/source/coco.py#L155-L163)

**Comment by Laughing-q** [comment]
@AyushExel 
- gt_bbox: with shape [N, 4], normlized.
- gt_class: with shape [N, 1].

**Comment by AyushExel** [comment]
Okay got it. I can attempt a fix in 2 hours

**Comment by Laughing-q** [comment]
@AyushExel or maybe we should consider migrating loss calculation into model just like other frameworks, then the workflow would be like this:
```python
batch = self.preprocess_batch(batch)
preds, loss = self.model(batch)
```
BTW following is our current workflow in trainer:
```python
batch = self.preprocess_batch(batch)
preds = self.model(batch['img'])
self.loss, self.loss_items = self.criterion(preds, batch)
```
I think this could be done since actually the loss part `criterion` in our trainer is only several lines of code. Migrating it into model could be done easily.
image

**Comment by AyushExel** [comment]
@Laughing-q Okay some progress. I used a very bad register solution to move ahead :) But I just want to make the trainer work and then we can make it elegant. Now, I'm here. I'll continue moving ahead after meetings but I think we might have something by tomorrow.
Screenshot 2023-05-22 at 7 50 46 PM

**Comment by AyushExel** [comment]
> @AyushExel or maybe we should consider migrating loss calculation into model just like other frameworks, then the workflow would be like this:
> 
> ```python
> batch = self.preprocess_batch(batch)
> preds, loss = self.model(batch)
> ```
> 
> BTW following is our current workflow in trainer:
> 
> ```python
> batch = self.preprocess_batch(batch)
> preds = self.model(batch['img'])
> self.loss, self.loss_items = self.criterion(preds, batch)
> ```
> 
> I think this could be done since actually the loss part `criterion` in our trainer is only several lines of code. Migrating it into model could be done easily. image

Agree.. But this would be a separate change right? Or do you think we should include it in this PR itself?

**Comment by Laughing-q** [comment]
@AyushExel actually I just feel like the current design look a bit odd when we take "img" out from `batch` first and produce `preds` then pass `preds` and `batch`(use batch again) to criterion. Why don't we just pass `batch` once(the first design) and get it all(preds and loss)?

**Comment by Laughing-q** [comment]
@AyushExel probably in this PR together? or we'll probably need to create a new trainer for rt-detr for now?

**Comment by AyushExel** [comment]
@Laughing-q okay. I think either way we need to traverse through the code execution and fix all the little tenor or paddle-to-torch-related typos like I'm doing with moving the trace where code fails.
Can you propose something if you want to include the change in this PR itself before you go to bed. I'll try to move ahead with that later tonight.
Anyway, I have to jump in meetings in 3 mins, I'll be back after 1.5 hours

**Comment by Laughing-q** [comment]
something like we move loss part into `DetectionModel` class and then we just need a new class named `RTDETR` inherited from `DetectionModel` with modifying the loss part to rt-detr loss.

**Comment by AyushExel** [comment]
Hey, it's working!!! No errors while training :)

**Comment by Laughing-q** [comment]
@AyushExel yes no errors for now, but still some TODOs:
- verify the correctness of training, now it runs no errors but we don't know if the training goes correctly.
- there actually would raise an error in bipartite graph matching process when the predictions are nan.
- the code is totally a mess, we need to completely clean it up.

cc @glenn-jocher

**Comment by AyushExel** [comment]
@Laughing-q  Yeah I guess we can first handle the register/loss in head part in a separate PR and pause this one here?

**Comment by Laughing-q** [comment]
@AyushExel sounds good!

**Comment by glenn-jocher** [comment]
@AyushExel @Laughing-q awesome guys. I can help clean up the code. I can try to train a model on VOC to compare to YOLOv8 quickly. Is the code ready for training?

**Comment by Laughing-q** [comment]
@glenn-jocher From now it runs with no errors but haven't verified the correctness of training yet. You can try to launch a training if you want. :)

**Comment by Kayzwer** [comment]
please review my pull request https://github.com/ultralytics/ultralytics/pull/2817, I fix the error based on message by pep8

**Comment by Laughing-q** [comment]
@Kayzwer Thank! This PR is still working on progress, your PR looks good I'll merge it in this branch.

**Comment by glenn-jocher** [comment]
@Kayzwer hi there, thank you very much for bringing the issue to our attention and providing a fix for it! We appreciate and value contributions from the community which help improve the reliability and performance of the YOLOv8 repository. Our maintainer will review your pull request and merge it if it aligns with our standards for quality and maintainability. Thank you again for your contribution, and please feel free to contribute any further improvements or bug fixes that you may come across in the future!

**Comment by Kayzwer** [comment]
fix unused import https://github.com/ultralytics/ultralytics/pull/2838

**Comment by glenn-jocher** [comment]
@Kayzwer thanks for bringing this to our attention! It looks like the code includes an unused import statement. Our team has reviewed your pull request and we appreciate your effort in fixing this issue. We'll take a closer look and consider merging it in the upcoming release. Please let us know if you have any further suggestions in improving the codebase.

**Comment by Kayzwer** [comment]
https://github.com/ultralytics/ultralytics/pull/2846

**Comment by AyushExel** [comment]
@Kayzwer Hey this PR is a WIP. We'll fix all linting and pre-commit errors before merging. This PR will change very quickly so I'd suggest holding off on making PRs to this branch until it's merged in main. Instead, if you'd like to make PRs on the main branch, that'd be more helpful

**Comment by Laughing-q** [comment]
@AyushExel @glenn-jocher ok this PR is ready to go I think.

**Comment by glenn-jocher** [comment]
@Laughing-q got it! Sorry didn't see this earlier.
